### PR TITLE
Code Release: Prime Sample Attention in Object Detection (CVPR 2020)

### LINKED ISSUE #1051: any plans for Prime Sample Attention in Object Detection?

### LINKED ISSUE #1749: When the source-code of PISA will be released?
I want to know that when the source-code of PISA will be released, thank you very much.

### PR DESCRIPTION
**Pull request by yhcao6**
Fix #1051, fix #1749

### CODE REVIEW

File: mmdet/core/bbox/samplers/score_hlr_sampler.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            num_rand = num_expected - num_hlr
+            if num_valid > 0:
+                valid_rois = rois[valid_inds]
+                valid_max_score = max_score[valid_inds]
+                valid_argmax_score = argmax_score[valid_inds]
+                valid_bbox_pred = bbox_pred[valid_inds]
+
+                valid_bbox_pred = valid_bbox_pred.view(
+                    valid_bbox_pred.size(0), -1, 4)
+                selected_bbox_pred = valid_bbox_pred[range(num_valid),
```
[the remark this answers is not in the document]
Comment (yhcao6) [reply mmdet/core/bbox/samplers/score_hlr_sampler.py:126]: valid_bbox_pred[:, valid_argmax_score] != valid_bbox_pred[range(num_valid), valid_argmax_score]

File: mmdet/core/bbox/samplers/score_hlr_sampler.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+
+@BBOX_SAMPLERS.register_module
+class ScoreHLRSampler(BaseSampler):
+    """Importance-based Sample Reweighting (ISR_N), negative part.
+
+    Args:
+        num (int): Sample total num.
+        pos_fraction (float): Fraction of positive.
+        context (obj): RoI head that the sampler belongs to.
```
[the remark this answers is not in the document]
Comment (yhcao6) [reply mmdet/core/bbox/samplers/score_hlr_sampler.py]: Now I rewrite it as context `(:obj: BaseRoIHead)`, I am not sure if it is correct.

**Comment by yhcao6** [comment]
> Hard work on this PR! The formatting is okay and I have tested faster-rcnn and retinanet and it is able to train. I have added several point-to-point comments below. There are also some general comments on this PR.
> 
> 1. More detailed docstring and comments are needed. It is less intuitive to understand the code without reading through the paper.
> 2. py-test functions of the added heads, samplers and nms_match ops should also be added.
> 3. (Maybe this should be a long-run process). Some of the basic functions e.g.methods in anchor_heads need to be massively modified as to its arguments and returning objects. I think a reconstruction of these fundamental functions is necessary to make them more flexible.

Thanks for the comments. I have modified according to the comments and also propose some problems. Could you check it again?

File: mmdet/models/dense_heads/anchor_head.py [hunk 3]
```diff
@@ -315,8 +316,8 @@ def get_targets(self,
             label_channels=label_channels,
             unmap_outputs=unmap_outputs)
         (all_labels, all_label_weights, all_bbox_targets, all_bbox_weights,
-         pos_inds_list, neg_inds_list) = results[:6]
-        rest_results = list(results[6:])  # user-added return values
+         pos_inds_list, neg_inds_list, sampling_results_list) = results[:7]
```
Comment (Johnson-Wang) [inline mmdet/models/dense_heads/anchor_head.py:319]: I am afraid the returning values of get_targets_single might be incompatible with that defined in [fsaf](https://github.com/open-mmlab/mmdetection/blob/bb8458dd4c2f2cd409253df8898b6fe4a91f2953/mmdet/models/dense_heads/fsaf_head.py#L43).

File: mmdet/models/dense_heads/pisa_retinanet_head.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+from mmdet.core import force_fp32, images_to_levels
+from ..builder import HEADS
+from ..losses import carl_loss, isr_p
+from .retina_head import RetinaHead
+
+
+@HEADS.register_module
+class PISARetinaHead(RetinaHead):
+
```
Comment (Johnson-Wang) [inline mmdet/models/dense_heads/pisa_retinanet_head.py:11]: docstring describing the major difference with the base class RetinaHead

File: mmdet/models/dense_heads/pisa_retinanet_head.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        num_level_anchors = [anchors.size(0) for anchors in anchor_list[0]]
+        # concat all level anchors and flags to a single tensor
+        concat_anchor_list = []
+        for i in range(len(anchor_list)):
+            concat_anchor_list.append(torch.cat(anchor_list[i]))
+        all_anchor_list = images_to_levels(concat_anchor_list,
+                                           num_level_anchors)
+
+        num_imgs = len(img_metas)
+        flatten_cls_scores = [
```
Comment (Johnson-Wang) [inline mmdet/models/dense_heads/pisa_retinanet_head.py:54]: Shall we define a new function dealing with the flattening and concatenation of score and bbox predictions? Concatenating predictions before the final loss calculation is quite common in other methods, and thus this function can be reusable.

File: mmdet/models/dense_heads/pisa_retinanet_head.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        flatten_labels = torch.cat(labels_list, dim=1).reshape(-1)
+        flatten_label_weights = torch.cat(
+            label_weights_list, dim=1).reshape(-1)
+        flatten_anchors = torch.cat(all_anchor_list, dim=1).reshape(-1, 4)
+        flatten_bbox_targets = torch.cat(
+            bbox_targets_list, dim=1).reshape(-1, 4)
+        flatten_bbox_weights = torch.cat(
+            bbox_weights_list, dim=1).reshape(-1, 4)
+
+        isr_cfg = self.train_cfg.get('isr', None)
```
Comment (Johnson-Wang) [inline mmdet/models/dense_heads/pisa_retinanet_head.py:76]: Add comments for explanation.

File: mmdet/models/losses/pisa_loss.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+          loss_cls,
+          bbox_coder,
+          k=2,
+          bias=0):
+    """Importance-based Sample Reweighting (ISR_P), positive part.
+
+    Args:
+        cls_score (Tensor): Predicted classification scores.
+        bbox_pred (Tensor): Predicted bbox deltas.
+        rois (Tensor): Anchors (single_stage) in shape (n, 4) or RoIs
```
Comment (Johnson-Wang) [inline mmdet/models/losses/pisa_loss.py:20]: missing bbox_targets args

File: mmdet/models/losses/pisa_loss.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        gt_i = sampling_results[i].pos_assigned_gt_inds
+        gts.append(gt_i + last_max_gt)
+        last_max_gt = gt_i.max() + 1
+    gts = torch.cat(gts)
+    assert len(gts) == num_pos
+
+    cls_score = cls_score.detach()
+    bbox_pred = bbox_pred.detach()
+
+    if rois.size(-1) == 5:
```
Comment (Johnson-Wang) [inline mmdet/models/losses/pisa_loss.py:55]: explanation

File: mmdet/models/losses/pisa_loss.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        pos_rois = rois[pos_label_inds][:, 1:]
+    else:
+        pos_rois = rois[pos_label_inds]
+
+    if bbox_pred.size(-1) > 4:
```
Comment (Johnson-Wang) [inline mmdet/models/losses/pisa_loss.py:60]: For every if - else block, we'd better add a remark explaining its reason.

File: mmdet/models/dense_heads/pisa_ssd_head.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                bbox_coder=self.bbox_coder,
+                **self.train_cfg.isr)
+            (new_labels, new_label_weights, new_bbox_targets,
+             new_bbox_weights) = all_targets
+            all_labels = new_labels.view(all_labels.shape)
+            all_label_weights = new_label_weights.view(all_label_weights.shape)
+            all_bbox_targets = new_bbox_targets.view(all_bbox_targets.shape)
+            all_bbox_weights = new_bbox_weights.view(all_bbox_weights.shape)
+
+        carl_loss_cfg = self.train_cfg.get('carl', None)
```
Comment (Johnson-Wang) [inline mmdet/models/dense_heads/pisa_ssd_head.py:85]: Better add the explanation of carl here

File: mmdet/models/roi_heads/pisa_roi_head.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
+from mmdet.core import bbox2roi
+from ..builder import HEADS
+from ..losses.pisa_loss import carl_loss, isr_p
+from .standard_roi_head import StandardRoIHead
+
+
+@HEADS.register_module
```
Comment (Johnson-Wang) [inline mmdet/models/roi_heads/pisa_roi_head.py:7]: register_module()

File: mmdet/models/dense_heads/pisa_retinanet_head.py (hunk 5 above)
Comment (yhcao6) [reply mmdet/models/dense_heads/pisa_retinanet_head.py:62]: I agree, but better to add in independent PR

File: mmdet/core/bbox/samplers/score_hlr_sampler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hellock) [reply mmdet/core/bbox/samplers/score_hlr_sampler.py]:
```
:obj:`BaseRoIHead`
```

File: mmdet/core/bbox/samplers/score_hlr_sampler.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+from ..builder import BBOX_SAMPLERS
+from ..transforms import bbox2roi
+from .base_sampler import BaseSampler
+from .sampling_result import SamplingResult
+
+
+@BBOX_SAMPLERS.register_module()
+class ScoreHLRSampler(BaseSampler):
+    """Importance-based Sample Reweighting (ISR_N), negative part.
+
```
Comment (hellock) [inline mmdet/core/bbox/samplers/score_hlr_sampler.py:13]: Add the paper link.

File: mmdet/core/bbox/samplers/score_hlr_sampler.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+    Score hierarchical local rank (HLR) differents with RandomSampler in
+    negative part. It firstly computes Score-HLR in a two-step way,
+    then linearly maps score hlr to the loss weights.
+
+    Args:
+        num (int): Sample total num.
+        pos_fraction (float): Fraction of positive.
```
Comment (hellock) [inline mmdet/core/bbox/samplers/score_hlr_sampler.py:20]: Fraction of positive samples.

File: mmdet/core/bbox/samplers/score_hlr_sampler.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+@BBOX_SAMPLERS.register_module()
+class ScoreHLRSampler(BaseSampler):
+    """Importance-based Sample Reweighting (ISR_N), negative part.
+
+    Score hierarchical local rank (HLR) differents with RandomSampler in
+    negative part. It firstly computes Score-HLR in a two-step way,
+    then linearly maps score hlr to the loss weights.
+
+    Args:
+        num (int): Sample total num.
```
Comment (hellock) [inline mmdet/core/bbox/samplers/score_hlr_sampler.py:19]: Total number of sampled RoIs.

File: mmdet/models/losses/pisa_loss.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+        bbox_coder (obj): BBox coder of the head.
+        k (float): Power of the non-linear mapping.
+        bias (float): Shift of the non-linear mapping.
+        avg_factor (int): Average factor used in regression loss.
+        sigmoid (bool): Activation of the classification score.
+
+    Return:
+        dict: CARL loss dict.
+    """
+    pos_label_inds = ((labels >= 0) & (labels < 80)).nonzero().reshape(-1)
```
Comment (Johnson-Wang) [inline mmdet/models/losses/pisa_loss.py:141]: 0<=labels<80 is hard-coded for COCO dataset. It's better to pass num_class to this function so that other datasets are also able to use it.

File: mmdet/core/bbox/samplers/score_hlr_sampler.py [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+                # valid_bbox_pred shape: [num_valid, #num_classes, 4]
+                valid_bbox_pred = valid_bbox_pred.view(
+                    valid_bbox_pred.size(0), -1, 4)
+                selected_bbox_pred = valid_bbox_pred[range(num_valid),
+                                                     valid_argmax_score]
+                pred_bboxes = self.bbox_head.bbox_coder.decode(
+                    valid_rois[:, 1:], selected_bbox_pred)
+                pred_bboxes_with_score = torch.cat(
+                    [pred_bboxes, valid_max_score[:, None]], -1)
+                group = nms_match(pred_bboxes_with_score, 0.5)
```
Comment (Johnson-Wang) [inline mmdet/core/bbox/samplers/score_hlr_sampler.py:173]: Is 0.5 universally used for all models? If not, we'd better pass an argument.

File: mmdet/models/losses/pisa_loss.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+        return bbox_targets
+
+    # merge pos_assigned_gt_inds of per image to a single tensor
+    gts = list()
+    last_max_gt = 0
+    for i in range(len(sampling_results)):
+        gt_i = sampling_results[i].pos_assigned_gt_inds
+        gts.append(gt_i + last_max_gt)
+        last_max_gt = gt_i.max() + 1
+    gts = torch.cat(gts)
```
Comment (Johnson-Wang) [inline mmdet/models/losses/pisa_loss.py:51]: Are there any chances that len(gts)==0? If yes, this line could cause an error.

File: mmdet/models/losses/pisa_loss.py [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+        pos_bbox_preds = bbox_pred[pos_label_inds, pos_labels]
+    else:
+        pos_bbox_preds = bbox_pred[pos_label_inds]
+    ori_loss_reg = loss_bbox(
+        pos_bbox_preds,
+        bbox_targets[pos_label_inds],
+        reduction_override='none') / avg_factor
+    carl_loss_weights = carl_loss_weights[:, None].expand(-1, 4)
+    loss_carl = (ori_loss_reg * carl_loss_weights).sum()
+    return dict(loss_carl=loss_carl[None])
```
Comment (Johnson-Wang) [inline mmdet/models/losses/pisa_loss.py:172]: The dimension expansion is unnecessary.

File: mmdet/core/bbox/samplers/score_hlr_sampler.py [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+        assigning results and ground truth bboxes.
+
+        Args:
+            assign_result (:obj:`AssignResult`): Bbox assigning results.
+            bboxes (Tensor): Boxes to be sampled from.
+            gt_bboxes (Tensor): Ground truth bboxes.
+            gt_labels (Tensor, optional): Class labels of ground truth bboxes.
+
+        Returns:
+            :obj:`SamplingResult`: Sampling result.
```
Comment (Johnson-Wang) [inline mmdet/core/bbox/samplers/score_hlr_sampler.py:232]: The description does not match the return.

File: mmdet/models/losses/pisa_loss.py (hunk 17 above)
Comment (yhcao6) [reply mmdet/models/losses/pisa_loss.py:51]: No chance. Since if `len(gts) == 0`, then `num(pos) == 0`, so it has already returned in line 44-45:
```
    if num_pos == 0:
        return bbox_targets

```

File: tests/test_nms.py [hunk 20]
```diff
@@ -68,3 +69,33 @@ def test_nms_device_and_dtypes_gpu():
...
+
+    # non empty tensor input
+    tensor_dets = torch.from_numpy(np_dets)
+    tensor_groups = nms_match(tensor_dets, iou_thr)
+    assert isinstance(tensor_groups[0], torch.Tensor)
+    for i in range(len(tensor_groups)):
+        assert np.equal(tensor_groups[i].numpy(), np_groups[i]).all()
+
+    # input of wrong shape
+    wrong_dets = np.zeros((2, 3))
```
Comment (Johnson-Wang) [inline tests/test_nms.py:99]: tests for size(0)==0

Comment (yhcao6) [reply tests/test_nms.py:99]: empty input has been tested in the first case.
